vectorstores/pgvector/engine.go

```diff
@@ -0,0 +1,141 @@
+	return getServiceAccountEmail(ctx)
+}
+
+type Column struct {
+	Name     string
+	DataType string
+	Nullable bool
+}
+
+func NewColumn(name, dataType string, nullable bool) (*Column, error) {
+	if name == "" {
+		return nil, errors.New("name should be provided")
+	}
+	if dataType == "" {
+		return nil, errors.New("dataType should be provided")
+	}
+	return &Column{Name: name, DataType: dataType, Nullable: nullable}, nil
+}
+
+type PostgresEngine struct {
+	Host                    string
+	Port                    int
+	Database                string
+	User                    string
+	Password                string
+	IAMAccountEmail         string
+	ServiceAccountRetriever ServiceAccountRetriever
+}
+
+// NewPostgresEngine initializes a new PostgresEngine
+func NewPostgresEngine(host, dbname, user, password, iamAccountEmail string, port int) *PostgresEngine {
+	return &PostgresEngine{
+		Host:            host,
+		Port:            port,
+		Database:        dbname,
+		User:            user,
+		Password:        password,
+		IAMAccountEmail: iamAccountEmail,
+	}
+}
+
+// CreatePostgresEngine establishes a connection to the PostgreSQL database.
+func (p *PostgresEngine) CreatePostgresEngine(ctx context.Context) (*pgx.Conn, error) {
+	err := p.assignUser(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error assigning user. Err: %s", err)
+	}
+
+	// Build the connection configuration
+	config, _ := pgx.ParseConfig("")
+	config.Host = p.Host
+	config.Port = uint16(p.Port)
+	config.Database = p.Database
+	config.User = p.User
+	config.Password = p.Password
+
+	// Make the connection to the PostgreSQL database
+	connection, err := pgx.ConnectConfig(ctx, config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to Database: %s, User: %s, Port: %v, Err: %s", p.Database, p.User, p.Port, err)
+	}
+	return connection, nil
+}
+
+func (p *PostgresEngine) assignUser(ctx context.Context) error {
+	// If both user and password are missing, use IAM-based authentication
+	if (p.User != "" && p.Password == "") || (p.User == "" && p.Password != "") {
+		return fmt.Errorf("only one of 'user' or 'password' were specified. Either both or none should be specified")
+	}
+	// If neither user nor password are provided, determine the IAM email
+	if p.User == "" && p.Password == "" {
+		// Set user's IAMAccountEmail if available
+		if p.IAMAccountEmail != "" {
+			p.User = p.IAMAccountEmail
+			return nil
+		}
+		if p.ServiceAccountRetriever != nil {
+			// Otherwise, attempt to retrueve the service account email
+			serviceAccountEmail, err := p.ServiceAccountRetriever.GetServiceAccountEmail(ctx)
+			if err != nil {
+				return fmt.Errorf("unable to retrieve service account email: %s", err)
+			}
+			p.User = serviceAccountEmail
+		}
+	}
+
+	// If no user can be determined, return an error.
+	if p.User == "" {
+		return errors.New("no valid user or IAM account email provided")
+	}
+
+	return nil
```

Comment:
I think this logic would be easier to follow if you check for valid conditions instead of invalid conditions
e.g. (also assuming returning "username" "usingIAMAuthN", and "err" to address my comment about functional return above):
```suggestion
        if (p.User == "" && p.Password == "") {
            // is the user doesn't provide credentials, automatically detect using IAM AuthN
            user, err := getServiceAccountEmail(ctx)
            if err !== nil {
                // handle err
            }
            return user, true, nil
        } else if (p.User != "" && p.Password != "") {
            //
            return user, false, nil
        }
        return nil, false, fmt.Errorf("only one of 'user' or 'password' were specified. Either both or none should be specified")
```